File: src/Koh.Emulator.App.Shared/Services/WebRtcLink.cs
using Koh.Emulator.Core.Serial;
using Microsoft.JSInterop;

namespace Koh.Emulator.App.Services;

/// <summary>
/// Prototype WebRTC-backed <see cref="ISerialLink"/>. Signaling is manual
/// paste-swap between the two instances — call <see cref="CreateOfferAsync"/>
/// on one peer, share the SDP string, call <see cref="AcceptOfferAsync"/> on
/// the other, share the answer back, and call <see cref="ApplyAnswerAsync"/>
/// on the first peer.
///
/// Caveats: <see cref="ExchangeByte"/> is synchronous but WebRTC is async.
/// The outgoing byte is queued for send (no blocking). The returned byte is
/// whatever is currently in the receive queue, falling back to $FF. For
/// real-time link-cable games that have tight handshakes, expect desync. The
/// data channel is fine for casual use (e.g. Pokémon trades which tolerate
/// lots of latency); tighter sync requires a deeper redesign.
/// </summary>
public sealed class WebRtcLink : ISerialLink, IAsyncDisposable
{
    private readonly IJSRuntime _js;
    private readonly DotNetObjectReference<WebRtcLink> _selfRef;

    public bool IsOpen { get; private set; }
    public event Action? Opened;
    public event Action? Closed;

    public WebRtcLink(IJSRuntime js)
    {
        _js = js;
        _selfRef = DotNetObjectReference.Create(this);
    }

    public async ValueTask RegisterAsync()
    {
        await _js.InvokeVoidAsync("kohWebRtcLink.register", _selfRef);
    }

    public ValueTask<string> CreateOfferAsync()
        => _js.InvokeAsync<string>("kohWebRtcLink.createOffer");

    public ValueTask<string> AcceptOfferAsync(string offerJson)
        => _js.InvokeAsync<string>("kohWebRtcLink.acceptOffer", offerJson);

    public ValueTask ApplyAnswerAsync(string answerJson)
        => _js.InvokeVoidAsync("kohWebRtcLink.applyAnswer", answerJson);

    public byte ExchangeByte(byte sent)
    {
        // Fire the send; don't await — the emulator's TickT is synchronous.
        _ = _js.InvokeVoidAsync("kohWebRtcLink.sendByte", sent);

        // Read whatever the peer has already delivered. Uses the synchronous
        // JS-interop fast path only when running in Blazor WASM; otherwise
        // returns $FF (no peer byte available yet).
        if (_js is IJSInProcessRuntime inProc)
        {
            int next = inProc.Invoke<int>("kohWebRtcLink.drainReceived");
            return next < 0 ? (byte)0xFF : (byte)next;
        }
        return 0xFF;
    }

    [JSInvokable]
    public void OnLinkOpened() { IsOpen = true; Opened?.Invoke(); }

    [JSInvokable]
    public void OnLinkClosed() { IsOpen = false; Closed?.Invoke(); }

    public async ValueTask DisposeAsync()
    {
        try { await _js.InvokeVoidAsync("kohWebRtcLink.close"); }
        catch { /* JS host may already be gone. */ }
        _selfRef.Dispose();
    }
}

File: src/Koh.Emulator.App.Shared/wwwroot/js/webrtc-link.js
// Prototype WebRTC data-channel transport for link-cable emulation.
// Signaling is manual paste-swap: one peer calls createOffer(), shares the
// returned SDP via out-of-band means (chat, etc.), the other calls
// acceptOffer() with that string and returns an answer, which the first peer
// feeds back through applyAnswer().
//
// The C# WebRtcLink drains the receive queue from ExchangeByte. No production
// use — no signaling server, no TURN, no reconnection handling.

window.kohWebRtcLink = (function () {
    let pc = null;
    let channel = null;
    let blazorRef = null;
    const receiveQueue = [];

    function onChannelOpen() {
        blazorRef?.invokeMethodAsync('OnLinkOpened');
    }

    function onChannelMessage(e) {
        const data = e.data instanceof ArrayBuffer ? new Uint8Array(e.data) : new Uint8Array(e.data.buffer ?? []);
        for (let i = 0; i < data.length; i++) receiveQueue.push(data[i]);
    }

    function wireChannel(ch) {
        channel = ch;
        channel.binaryType = 'arraybuffer';
        channel.onopen = onChannelOpen;
        channel.onmessage = onChannelMessage;
        channel.onclose = () => blazorRef?.invokeMethodAsync('OnLinkClosed');
    }

    return {
        register: function (dotNetRef) { blazorRef = dotNetRef; },

        createOffer: async function () {
            pc = new RTCPeerConnection();
            wireChannel(pc.createDataChannel('koh-serial', { ordered: true }));
            const offer = await pc.createOffer();
            await pc.setLocalDescription(offer);
            // Wait for ICE gathering so the SDP is complete before we hand it off.
            await new Promise(resolve => {
                if (pc.iceGatheringState === 'complete') { resolve(); return; }
                pc.addEventListener('icegatheringstatechange', () => {
                    if (pc.iceGatheringState === 'complete') resolve();
                });
            });
            return JSON.stringify(pc.localDescription);
        },

        acceptOffer: async function (offerJson) {
            pc = new RTCPeerConnection();
            pc.ondatachannel = (e) => wireChannel(e.channel);
            await pc.setRemoteDescription(JSON.parse(offerJson));
            const answer = await pc.createAnswer();
            await pc.setLocalDescription(answer);
            await new Promise(resolve => {
                if (pc.iceGatheringState === 'complete') { resolve(); return; }
                pc.addEventListener('icegatheringstatechange', () => {
                    if (pc.iceGatheringState === 'complete') resolve();
                });
            });
            return JSON.stringify(pc.localDescription);
        },

        applyAnswer: async function (answerJson) {
            if (!pc) throw new Error('call createOffer first');
            await pc.setRemoteDescription(JSON.parse(answerJson));
        },

        sendByte: function (byte) {
            if (channel?.readyState === 'open') {
                channel.send(new Uint8Array([byte]).buffer);
            }
        },

        drainReceived: function () {
            if (receiveQueue.length === 0) return -1;
            return receiveQueue.shift();
        },

        isOpen: function () { return channel?.readyState === 'open'; },

        close: function () {
            channel?.close();
            pc?.close();
            channel = null;
            pc = null;
            receiveQueue.length = 0;
        },
    };
})();
